pkg/updater/repository/repository.go

```diff
@@ -0,0 +1,324 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+// Package repository contains the packaging logic of the updater.
+package repository
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+)
+
+const (
+	previousVersionLink   = "previous"
+	stableVersionLink     = "stable"
+	experimentVersionLink = "experiment"
+)
+
+// Repository contains the stable and experimental package of a single artifact managed by the updater.
+//
+// On disk the repository is structured as follows:
+// .
+// ├── 7.50.0
+// ├── 7.51.0
+// ├── stable -> 7.50.0 (symlink)
+// └── experiment -> 7.51.0 (symlink)
+//
+// We voluntarily do not load the state of the repository in memory to avoid any bugs where
+// what's on disk and what's in memory are not in sync.
+// All the functions of the repository are "atomic" and ensure no invalid state can be reached
+// even if an error happens during their execution.
+// It is possible to end up with garbage left on disk if an error happens during some operations. This
+// is cleaned up during the next operation.
+type Repository struct {
+	RootPath string
+}
+
+// Status is the status of the repository.
+type Status struct {
+	Stable     string
+	Experiment string
+}
+
+// HasStable returns true if the repository has a stable package.
+func (s *Status) HasStable() bool {
+	return s.Stable != ""
+}
+
+// HasExperiment returns true if the repository has an experiment package.
+func (s *Status) HasExperiment() bool {
+	return s.Experiment != ""
+}
+
+// GetStatus returns the status of the repository.
+func (r *Repository) GetStatus() (*Status, error) {
+	repository, err := openRepository(r.RootPath)
+	if err != nil {
+		return nil, err
+	}
+	return &Status{
+		Stable:     repository.stable.Target(),
+		Experiment: repository.experiment.Target(),
+	}, nil
+}
+
+// Create creates a fresh new repository at the given root path
+// and moves the given stable source path to the repository as the first stable.
+// If a repository already exists at the given path, it is fully removed.
+//
+// 1. Remove the previous repository if it exists.
+// 2. Create the root directory.
+// 3. Move the stable source to the repository.
+// 4. Create the stable link.
+func (r *Repository) Create(name string, stableSourcePath string) error {
```

Comment:
That repository would not be used afterward tho because we only `Create` once when the very first version is installed with the `bootstrap` command. Having `Create` init the repository internally and not return it is an option but I don't think it brings a lot. wdyt ?